Source/Urho3D/CSharp/Managed/Container/SharedPtr.cs

```diff
@@ -0,0 +1,146 @@
+                Trace.WriteLine(
+                    $"Object of type {typeof(T).FullName} wasn't properly disposed. Please call Dispose on all SharedPtr<{typeof(T).Name}> instances.");
+        }
+
+        /// <summary>
+        /// Copy SharedPtr.
+        /// </summary>
+        /// <param name="other">SharedPtr instance to copy</param>
+        public SharedPtr(SharedPtr<T> other) : this(other?._value)
+        {
+        }
+
+        /// <summary>Indicates whether the current object is equal to another object of the same type.</summary>
+        /// <param name="other">An object to compare with this object.</param>
+        /// <returns>true if the current object is equal to the <paramref name="other">other</paramref> parameter; otherwise, false.</returns>
+        public bool Equals(SharedPtr<T> other)
+        {
+            if (ReferenceEquals(null, other)) return false;
+            if (ReferenceEquals(this, other)) return true;
+            return EqualityComparer<T>.Default.Equals(_value, other._value);
+        }
+
+        /// <summary>Determines whether the specified object is equal to the current object.</summary>
+        /// <param name="obj">The object to compare with the current object.</param>
+        /// <returns>true if the specified object  is equal to the current object; otherwise, false.</returns>
+        public override bool Equals(object obj)
+        {
+            if (ReferenceEquals(null, obj)) return false;
+            if (ReferenceEquals(this, obj)) return true;
+            if (obj.GetType() != this.GetType()) return false;
+            return Equals((SharedPtr<T>)obj);
+        }
+
+        /// <summary>Serves as the default hash function.</summary>
+        /// <returns>A hash code for the current object.</returns>
+        public override int GetHashCode()
+        {
+            return EqualityComparer<T>.Default.GetHashCode(_value);
+        }
+
+        /// <summary>
+        /// Implicitly convert SharedPtr into value.
+        /// </summary>
+        /// <param name="other"></param>
+        public static implicit operator T(SharedPtr<T> @this) => @this._value;
+
+        /// <summary>Indicates whether the two objects are equal.</summary>
+        public static bool operator ==(SharedPtr<T> left, SharedPtr<T> right)
+        {
+            return Equals(left, right);
+        }
+
+        /// <summary>Indicates whether the two objects are not equal.</summary>
+        public static bool operator !=(SharedPtr<T> left, SharedPtr<T> right)
+        {
+            return !Equals(left, right);
+        }
+
+        /// <summary>
+        /// Get the underlying value.
+        /// </summary>
+        public T Ptr => _value;
+
+        /// <summary>Performs application-defined tasks associated with freeing, releasing, or resetting unmanaged resources.</summary>
+        public void Dispose()
+        {
+            if (_value != null)
+            {
+                _value.ReleaseRef();
```

Comment:
This should set `_value` to `null`. In theory we can call `Dispose()` twice, then things blow up. After object is disposed it should no longer be functional.

Reply:
Setting value to null would change the hash of the object. Would it be ok?

Reply:
I think it should be fine. If such `SharedPtr<>` is stored in a dict as a key, then first it has to be removed, then disposed, otherwise operation makes no sense. If we are doing a lookup by `SharedPtr<>` and it is now empty, then it makes sense to look for a key that is also empty `SharedPtr<>`.

Reply:
No reason to waste time on equality methods now if they aren't in use. I'm going to delete this.